# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Parent  c7f2a65f635cfb690bb9f3bc3e1e1bde421a7e79
Bug 2071585 - Fix broken offline for POP.

We match the "silent" behaviour as in NntpClient.sys.mjs.
_getMail() now matches getNewNews().

diff --git a/mailnews/local/src/Pop3Client.sys.mjs b/mailnews/local/src/Pop3Client.sys.mjs
--- a/mailnews/local/src/Pop3Client.sys.mjs
+++ b/mailnews/local/src/Pop3Client.sys.mjs
@@ -412,12 +412,13 @@ export class Pop3Client {
       // If there's a message window on the URI, then we should alert the user.
       // Otherwise (i.e. if the getter for `msgWindow` raised
       // `NS_ERROR_NULL_POINTER`), this is a background operation and we should
       // tell the mail session to only call the listeners but not alert.
-      let silent = false;
+      let silent = true;
       try {
-        this.runningUri.msgWindow;
-        silent = false;
+        if (this.runningUri.msgWindow) {
+          silent = false;
+        }
       } catch (ex) {
         if (
           !(ex instanceof Ci.nsIException) &&
           ex.result != Cr.NS_ERROR_NULL_POINTER
@@ -923,12 +924,13 @@ export class Pop3Client {
       // If there's a message window on the URI, then we should alert the user.
       // Otherwise (i.e. if the getter for `msgWindow` raised
       // `NS_ERROR_NULL_POINTER`), this is a background operation and we should
       // tell the mail session to only call the listeners but not alert.
-      let silent = false;
+      let silent = true;
       try {
-        this.runningUri.msgWindow;
-        silent = false;
+        if (this.runningUri.msgWindow) {
+          silent = false;
+        }
       } catch (ex) {
         if (
           !(ex instanceof Ci.nsIException) &&
           ex.result != Cr.NS_ERROR_NULL_POINTER
diff --git a/mailnews/local/src/Pop3IncomingServer.sys.mjs b/mailnews/local/src/Pop3IncomingServer.sys.mjs
--- a/mailnews/local/src/Pop3IncomingServer.sys.mjs
+++ b/mailnews/local/src/Pop3IncomingServer.sys.mjs
@@ -114,8 +114,11 @@ export class Pop3IncomingServer extends 
     return MailServices.pop3.verifyLogon(this, urlListener, msgWindow);
   }
 
   performBiff(msgWindow) {
+    if (Services.io.offline) {
+      return;
+    }
     this.performingBiff = true;
     const inbox = this.rootMsgFolder.getFolderWithFlags(
       Ci.nsMsgFolderFlags.Inbox
     );
diff --git a/mailnews/local/src/Pop3Service.sys.mjs b/mailnews/local/src/Pop3Service.sys.mjs
--- a/mailnews/local/src/Pop3Service.sys.mjs
+++ b/mailnews/local/src/Pop3Service.sys.mjs
@@ -3,8 +3,11 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 import { Pop3Client } from "resource:///modules/Pop3Client.sys.mjs";
 
+// See NS_MSG_ERROR_OFFLINE in nsMsgUtils.h.
+const NS_MSG_ERROR_OFFLINE = 0x80550014;
+
 /**
  * @implements {nsIPop3Service}
  */
 export class Pop3Service {
@@ -22,8 +25,14 @@ export class Pop3Service {
     this._getMail(false, msgWindow, urlListener, inbox, server);
   }
 
   verifyLogon(server, urlListener, msgWindow) {
+    if (Services.io.offline) {
+      throw Components.Exception(
+        "Cannot verify logon while offline",
+        NS_MSG_ERROR_OFFLINE
+      );
+    }
     const client = new Pop3Client(server);
     client.urlListener = urlListener;
     client.connect();
     client.onOpen = () => {
@@ -58,8 +67,14 @@ export class Pop3Service {
     }
   }
 
   _getMail(downloadNewMail, msgWindow, urlListener, inbox, server) {
+    if (Services.io.offline) {
+      throw Components.Exception(
+        "Cannot get mail while offline",
+        NS_MSG_ERROR_OFFLINE
+      );
+    }
     server.wrappedJSObject.withClient(client => {
       client.runningUri.msgWindow = msgWindow;
       client.urlListener = urlListener;
       client.onOpen = () => {
